### PR TITLE
DynamoDB support

### PR DESCRIPTION
**Pull request by saratlingamarla**
Before submitting pull requests, please see the
[Development documentation](http://awslimitchecker.readthedocs.org/en/latest/development.html)
and specifically the [Pull Request Guidelines](http://awslimitchecker.readthedocs.org/en/latest/development.html#pull-requests).

__IMPORTANT:__ Please take note of the below checklist, especially the first two items.

# Pull Request Checklist

- [x] All pull requests should be __against the develop branch__, not master.
- [x] All pull requests must include the Contributor License Agreement (see below).
- [x] Code should conform to the [Development Guidelines](http://awslimitchecker.readthedocs.org/en/latest/development.html#guidelines):
    - [x] pep8 compliant with some exceptions (see pytest.ini)
    - [ ] 100% test coverage with pytest (with valid tests). If you have difficulty
      writing tests for the code, feel free to ask for help or submit the PR without tests.
    - [x] Complete, correctly-formatted documentation for all classes, functions and methods.
    - [x] documentation has been rebuilt with ``tox -e docs``
    - [x] Connections to the AWS services should only be made by the class's
      ``connect()`` and ``connect_resource()`` methods, inherited from
      [awslimitchecker.connectable.Connectable](http://awslimitchecker.readthedocs.org/en/latest/awslimitchecker.connectable.html)
    - [x] All modules should have (and use) module-level loggers.
    - [x] **Commit messages** should be meaningful, and reference the Issue number
      if you're working on a GitHub issue (i.e. "issue #x - <message>"). Please
      refrain from using the "fixes #x" notation unless you are *sure* that the
      the issue is fixed in that commit.
    - [x] Git history is fully intact; please do not squash or rewrite history.

## Contributor License Agreement

By submitting this work for inclusion in awslimitchecker, I agree to the following terms:

* The contribution included in this request (and any subsequent revisions or versions of it)
  is being made under the same license as the awslimitchecker project (the Affero GPL v3,
  or any subsequent version of that license if adopted by awslimitchecker).
* My contribution may perpetually be included in and distributed with awslimitchecker; submitting
  this pull request grants a perpetual, global, unlimited license for it to be used and distributed
  under the terms of awslimitchecker's license.
* I have the legal power and rights to agree to these terms.

### CODE REVIEW

**Comment by saratlingamarla** [comment]
I'm getting an AttributeError, not sure why -

region_name = self.conn._client_config.region_name
AttributeError: '_DynamodbService' object has no attribute 'conn'

File: awslimitchecker/services/dynamodb.py [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+
+    def get_limits(self):
+        """
+        Return all known limits for this service, as a dict of their names
+        to :py:class:`~.AwsLimit` objects.
+
+        :returns: dict of limit names to :py:class:`~.AwsLimit` objects
+        :rtype: dict
+        """
+        region_name = self.conn._client_config.region_name
```
Comment (jantman) [inline awslimitchecker/services/dynamodb.py]: You need a call to `self.connect()` before this line.

**Comment by jantman** [comment]
Sorry I haven't had a chance to look at this yet, it might be another day or two before I can.

That AttributeError is also causing [the unit tests to fail](https://travis-ci.org/jantman/awslimitchecker/builds/274013112?utm_source=github_status&utm_medium=notification).

The AttributeError is because ``self.get_limits()`` is missing a call to ``self.connect()``.

**Comment by saratlingamarla** [comment]
adding a self.connect() in get_limits() still gives me the same Error.

File: awslimitchecker/services/dynamodb.py (hunk 1 above)
Comment (saratlingamarla) [reply awslimitchecker/services/dynamodb.py]: that's exactly what I did

Comment (jantman) [reply awslimitchecker/services/dynamodb.py]: Ok. Can you then try to post the full traceback, and debug-level output?

**Comment by jantman** [comment]
@saratlingamarla I should have some time in the next few days to finally work on the next release. It's up to you if you'd like me to assist with this PR however I can, or just pick it up where it is right now and finish it up myself.

If you'd like some help debugging the AttributeError, please (1) make sure the code pushed up in your branch is updated with what you're running, and (2) paste the full debug-level (``-vv``) output and the full traceback.

**Comment by saratlingamarla** [comment]
@jantman please take a look at the full debug level output :
 
(awslimitchecker) sarat@sarat-Inspiron-3542:~$ awslimitchecker -vv
awslimitchecker 0.11.0 is AGPL-licensed free software; all users have a right to the full source code of this version. See <https://github.com/jantman/awslimitchecker>
2017-09-13 23:05:02,220 [DEBUG checker.py:200 - awslimitchecker.checker._boto_conn_kwargs() ] Connecting to region None
Traceback (most recent call last):
  File "/home/sarat/Desktop/awslimitchecker/bin/awslimitchecker", line 11, in <module>
    load_entry_point('awslimitchecker==0.11.0', 'console_scripts', 'awslimitchecker')()
  File "/home/sarat/Desktop/awslimitchecker/local/lib/python2.7/site-packages/awslimitchecker/runner.py", line 405, in console_entry_point
    r.console_entry_point()
  File "/home/sarat/Desktop/awslimitchecker/local/lib/python2.7/site-packages/awslimitchecker/runner.py", line 362, in console_entry_point
    ta_refresh_timeout=args.ta_refresh_timeout
  File "/home/sarat/Desktop/awslimitchecker/local/lib/python2.7/site-packages/awslimitchecker/checker.py", line 157, in __init__
    boto_conn_kwargs)
  File "/home/sarat/Desktop/awslimitchecker/local/lib/python2.7/site-packages/awslimitchecker/services/base.py", line 104, in __init__
    self.limits = self.get_limits()
  File "/home/sarat/Desktop/awslimitchecker/local/lib/python2.7/site-packages/awslimitchecker/services/dynamodb.py", line 149, in get_limits
    self.connect()
  File "/home/sarat/Desktop/awslimitchecker/local/lib/python2.7/site-packages/awslimitchecker/connectable.py", line 85, in connect
    if self.conn is not None:
AttributeError: '_DynamodbService' object has no attribute 'conn'

**Comment by jantman** [comment]
Ok, it took me a while to figure out exactly what was going on here.

This is the only service that has per-region limits, so this is the only service that's calling ``self.connect()`` in ``get_limits()``. It turns out there's a slight ordering issue that was causing this to break.

In [awslimitchecker/services/base.py __init__()](https://github.com/jantman/awslimitchecker/blob/3468e28ac063aae70e4c8d0bc17e5d17a3bf5c34/awslimitchecker/services/base.py#L103) you'll need to switch lines 103-4 and 105-6 around so that ``self.conn`` and ``self.resource_conn`` are created before ``get_limits()`` is called. I.e. change this:

```python
        self.limits = {}
        self.limits = self.get_limits()
        self.conn = None
        self.resource_conn = None
```

to:

```python
        self.conn = None
        self.resource_conn = None
        self.limits = {}
        self.limits = self.get_limits()
```

That should at least get the code executing.

**Comment by saratlingamarla** [comment]
debug output -

(awslimitchecker) sarat@sarat-Inspiron-3542:~/Desktop/awslimitchecker$ awslimitchecker -vv
awslimitchecker 0.11.0@bf4fa7dc is AGPL-licensed free software; all users have a right to the full source code of this version. See <https://github.com/jantman/awslimitchecker>
2017-09-15 22:46:02,011 [DEBUG checker.py:200 - awslimitchecker.checker._boto_conn_kwargs() ] Connecting to region None
Traceback (most recent call last):
  File "/home/sarat/Desktop/awslimitchecker/bin/awslimitchecker", line 11, in <module>
    load_entry_point('awslimitchecker', 'console_scripts', 'awslimitchecker')()
  File "/home/sarat/Desktop/awslimitchecker/src/awslimitchecker/awslimitchecker/runner.py", line 405, in console_entry_point
    r.console_entry_point()
  File "/home/sarat/Desktop/awslimitchecker/src/awslimitchecker/awslimitchecker/runner.py", line 362, in console_entry_point
    ta_refresh_timeout=args.ta_refresh_timeout
  File "/home/sarat/Desktop/awslimitchecker/src/awslimitchecker/awslimitchecker/checker.py", line 157, in __init__
    boto_conn_kwargs)
  File "/home/sarat/Desktop/awslimitchecker/src/awslimitchecker/awslimitchecker/services/base.py", line 106, in __init__
    self.limits = self.get_limits()
  File "/home/sarat/Desktop/awslimitchecker/src/awslimitchecker/awslimitchecker/services/dynamodb.py", line 149, in get_limits
    self.connect()
  File "/home/sarat/Desktop/awslimitchecker/src/awslimitchecker/awslimitchecker/connectable.py", line 88, in connect
    self.conn = boto3.client(self.api_name, **kwargs)
  File "/home/sarat/Desktop/awslimitchecker/local/lib/python2.7/site-packages/boto3/__init__.py", line 83, in client
    return _get_default_session().client(*args, **kwargs)
  File "/home/sarat/Desktop/awslimitchecker/local/lib/python2.7/site-packages/boto3/session.py", line 263, in client
    aws_session_token=aws_session_token, config=config)
  File "/home/sarat/Desktop/awslimitchecker/local/lib/python2.7/site-packages/botocore/session.py", line 836, in create_client
    client_config=config, api_version=api_version)
  File "/home/sarat/Desktop/awslimitchecker/local/lib/python2.7/site-packages/botocore/client.py", line 71, in create_client
    verify, credentials, scoped_config, client_config, endpoint_bridge)
  File "/home/sarat/Desktop/awslimitchecker/local/lib/python2.7/site-packages/botocore/client.py", line 283, in _get_client_args
    verify, credentials, scoped_config, client_config, endpoint_bridge)
  File "/home/sarat/Desktop/awslimitchecker/local/lib/python2.7/site-packages/botocore/args.py", line 45, in get_client_args
    endpoint_url, is_secure, scoped_config)
  File "/home/sarat/Desktop/awslimitchecker/local/lib/python2.7/site-packages/botocore/args.py", line 111, in compute_client_args
    service_name, region_name, endpoint_url, is_secure)
  File "/home/sarat/Desktop/awslimitchecker/local/lib/python2.7/site-packages/botocore/client.py", line 356, in resolve
    service_name, region_name)
  File "/home/sarat/Desktop/awslimitchecker/local/lib/python2.7/site-packages/botocore/regions.py", line 122, in construct_endpoint
    partition, service_name, region_name)
  File "/home/sarat/Desktop/awslimitchecker/local/lib/python2.7/site-packages/botocore/regions.py", line 135, in _endpoint_for_partition
    raise NoRegionError()
botocore.exceptions.NoRegionError: You must specify a region.

**Comment by jantman** [comment]
Ok, thanks. That looks better.

Do you think you could run it so the code actually executes, i.e. with your AWS credentials set in the environment? I assume if you're writing this for your own need, you'd want to know that the limit and usage information is accurate...

**Comment by saratlingamarla** [comment]
Yes, sorry about the delay, but I verified the usages and they look right:

(awslimitchecker) sarat@sarat-Inspiron-3542:~/Desktop/awslimitchecker$ awslimitchecker -u
awslimitchecker 0.11.0@97d348c0 is AGPL-licensed free software; all users have a right to the full source code of this version. See <https://github.com/jantman/awslimitchecker>
WARNING:awslimitchecker.trustedadvisor:Cannot check TrustedAdvisor: AWS Premium Support Subscription is required to use this service.
AutoScaling/Auto Scaling groups                        0
AutoScaling/Launch configurations                      0
CloudFormation/Stacks                                  1
Dynamodb/Global Secondary Indexes                      max: arn:aws:dynamodb:us-east-1:339518693606:table/test2=2 (arn:aws:dynamodb:us-east-1:339518693606:table/test1=0, arn:aws:dynamodb:us-east-1:339518693606:table/test2=2)
Dynamodb/Local Secondary Indexes                       max: arn:aws:dynamodb:us-east-1:339518693606:table/test1=0 (arn:aws:dynamodb:us-east-1:339518693606:table/test1=0, arn:aws:dynamodb:us-east-1:339518693606:table/test2=0)
Dynamodb/Read Capacity (region)                        20
Dynamodb/Read Capacity (table)                         max: arn:aws:dynamodb:us-east-1:339518693606:table/test2=15 (arn:aws:dynamodb:us-east-1:339518693606:table/test1=5, arn:aws:dynamodb:us-east-1:339518693606:table/test2=15)
Dynamodb/Table Count (region)                          2
Dynamodb/Write Capacity (region)                       20
Dynamodb/Write Capacity (table)                        max: arn:aws:dynamodb:us-east-1:339518693606:table/test2=15 (arn:aws:dynamodb:us-east-1:339518693606:table/test1=5, arn:aws:dynamodb:us-east-1:339518693606:table/test2=15)
EBS/Active snapshots                                   0
EBS/Active volumes                                     0
EBS/Cold (HDD) volume storage (GiB)                    0
EBS/General Purpose (SSD) volume storage (GiB)         0
EBS/Magnetic volume storage (GiB)                      0
EBS/Provisioned IOPS                                   0
EBS/Provisioned IOPS (SSD) storage (GiB)               0
EBS/Throughput Optimized (HDD) volume storage (GiB)    0
EC2/Elastic IP addresses (EIPs)                        0
EC2/Max active spot fleets per region                  0
EC2/Max launch specifications per spot fleet           <unknown>
EC2/Max spot instance requests per region              0
EC2/Max target capacity for all spot fleets in region  0
EC2/Max target capacity per spot fleet                 <unknown>
EC2/Rules per VPC security group                       max: sg-bd1ec5cd=1 (sg-bd1ec5cd=1, sg-04823d74=1)
EC2/Running On-Demand EC2 instances                    0
EC2/Running On-Demand c1.medium instances              <unknown>
EC2/Running On-Demand c1.xlarge instances              <unknown>
EC2/Running On-Demand c3.2xlarge instances             <unknown>

**Comment by jantman** [comment]
Ok, thanks so much for this! I'm going to be working on finishing up the tests for this either today or tomorrow; when I do, I'll push the fixed tests to your branch and then merge this.

**Pull request merged**